File: binding/src/libqabstractfileiconprovider.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qabstractfileiconprovider.html
pub const qabstractfileiconprovider = struct {

    /// New constructs a new QAbstractFileIconProvider object.
    pub fn New() QtC.QAbstractFileIconProvider {
        return qtc.QAbstractFileIconProvider_new();
    }


    pub fn Icon(self: ?*anyopaque, param1: i32) QtC.QIcon {
        return qtc.QAbstractFileIconProvider_Icon(@ptrCast(self), @intCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnIcon(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) QtC.QIcon) void {
        qtc.QAbstractFileIconProvider_OnIcon(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseIcon(self: ?*anyopaque, param1: i32) QtC.QIcon {
        return qtc.QAbstractFileIconProvider_QBaseIcon(@ptrCast(self), @intCast(param1));
    }

    pub fn Icon2(self: ?*anyopaque, param1: ?*anyopaque) QtC.QIcon {
        return qtc.QAbstractFileIconProvider_Icon2(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnIcon2(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) QtC.QIcon) void {
        qtc.QAbstractFileIconProvider_OnIcon2(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseIcon2(self: ?*anyopaque, param1: ?*anyopaque) QtC.QIcon {
        return qtc.QAbstractFileIconProvider_QBaseIcon2(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Type(self: ?*anyopaque, param1: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAbstractFileIconProvider_Type(@ptrCast(self), @ptrCast(param1));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qabstractfileiconprovider.Type: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QAbstractFileIconProvider_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, param1: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAbstractFileIconProvider_QBaseType(@ptrCast(self), @ptrCast(param1));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qabstractfileiconprovider.Type: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetOptions(self: ?*anyopaque, options: i32) void {
        qtc.QAbstractFileIconProvider_SetOptions(@ptrCast(self), @intCast(options));
    }

    /// Allows for overriding the related default method
    pub fn OnSetOptions(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QAbstractFileIconProvider_OnSetOptions(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetOptions(self: ?*anyopaque, options: i32) void {
        qtc.QAbstractFileIconProvider_QBaseSetOptions(@ptrCast(self), @intCast(options));
    }

    pub fn Options(self: ?*anyopaque, ) i32 {
        return qtc.QAbstractFileIconProvider_Options(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnOptions(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QAbstractFileIconProvider_OnOptions(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseOptions(self: ?*anyopaque, ) i32 {
        return qtc.QAbstractFileIconProvider_QBaseOptions(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAbstractFileIconProvider_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/abstractfileiconprovider.html#types
pub const enums = struct {
    pub const IconType = enum {
        pub const Computer: i32 = 0;
        pub const Desktop: i32 = 1;
        pub const Trashcan: i32 = 2;
        pub const Network: i32 = 3;
        pub const Drive: i32 = 4;
        pub const Folder: i32 = 5;
        pub const File: i32 = 6;
    };

    pub const Option = enum {
        pub const DontUseCustomDirectoryIcons: i32 = 1;
    };

};

File: binding/src/libqabstractfileiconprovider.cpp
#include <QAbstractFileIconProvider>
#include <QFileInfo>
#include <QIcon>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qabstractfileiconprovider.h>
#include "libqabstractfileiconprovider.h"
#include "libqabstractfileiconprovider.hxx"

QAbstractFileIconProvider* QAbstractFileIconProvider_new() {
	 return new VirtualQAbstractFileIconProvider();
}

QIcon* QAbstractFileIconProvider_Icon(const QAbstractFileIconProvider* self, int param1) {
	return new QIcon(self->icon(static_cast<QAbstractFileIconProvider::IconType>(param1)));
}

QIcon* QAbstractFileIconProvider_Icon2(const QAbstractFileIconProvider* self, const QFileInfo* param1) {
	return new QIcon(self->icon(*param1));
}

libqt_string QAbstractFileIconProvider_Type(const QAbstractFileIconProvider* self, const QFileInfo* param1) {
	QString _ret = self->type(*param1);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QAbstractFileIconProvider_SetOptions(QAbstractFileIconProvider* self, int options) {
	self->setOptions(static_cast<QFlags<QAbstractFileIconProvider::Option>>(options));
}

int QAbstractFileIconProvider_Options(const QAbstractFileIconProvider* self) {
	return self->options();
}

// Base class handler implementation
QIcon* QAbstractFileIconProvider_QBaseIcon(const QAbstractFileIconProvider* self, int param1) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Icon_IsBase(true);
	return new QIcon(vqabstractfileiconprovider->icon(static_cast<QAbstractFileIconProvider::IconType>(param1)));
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractFileIconProvider_OnIcon(const QAbstractFileIconProvider* self, intptr_t slot) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Icon_Callback(reinterpret_cast<VirtualQAbstractFileIconProvider::QAbstractFileIconProvider_Icon_Callback>(slot));
}
}

// Base class handler implementation
QIcon* QAbstractFileIconProvider_QBaseIcon2(const QAbstractFileIconProvider* self, const QFileInfo* param1) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Icon2_IsBase(true);
	return new QIcon(vqabstractfileiconprovider->icon(*param1));
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractFileIconProvider_OnIcon2(const QAbstractFileIconProvider* self, intptr_t slot) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Icon2_Callback(reinterpret_cast<VirtualQAbstractFileIconProvider::QAbstractFileIconProvider_Icon2_Callback>(slot));
}
}

// Base class handler implementation
libqt_string QAbstractFileIconProvider_QBaseType(const QAbstractFileIconProvider* self, const QFileInfo* param1) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Type_IsBase(true);
	QString _ret = vqabstractfileiconprovider->type(*param1);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractFileIconProvider_OnType(const QAbstractFileIconProvider* self, intptr_t slot) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Type_Callback(reinterpret_cast<VirtualQAbstractFileIconProvider::QAbstractFileIconProvider_Type_Callback>(slot));
}
}

// Base class handler implementation
void QAbstractFileIconProvider_QBaseSetOptions(QAbstractFileIconProvider* self, int options) {
	auto* vqabstractfileiconprovider = dynamic_cast<VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_SetOptions_IsBase(true);
	vqabstractfileiconprovider->setOptions(static_cast<QFlags<QAbstractFileIconProvider::Option>>(options));
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractFileIconProvider_OnSetOptions(QAbstractFileIconProvider* self, intptr_t slot) {
	auto* vqabstractfileiconprovider = dynamic_cast<VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_SetOptions_Callback(reinterpret_cast<VirtualQAbstractFileIconProvider::QAbstractFileIconProvider_SetOptions_Callback>(slot));
}
}

// Base class handler implementation
int QAbstractFileIconProvider_QBaseOptions(const QAbstractFileIconProvider* self) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Options_IsBase(true);
	return vqabstractfileiconprovider->options();
}
}

// Auxiliary method to allow providing re-implementation
void QAbstractFileIconProvider_OnOptions(const QAbstractFileIconProvider* self, intptr_t slot) {
	auto* vqabstractfileiconprovider = dynamic_cast<const VirtualQAbstractFileIconProvider*>(self);
	if (vqabstractfileiconprovider && vqabstractfileiconprovider->isVirtualQAbstractFileIconProvider) {
vqabstractfileiconprovider->setQAbstractFileIconProvider_Options_Callback(reinterpret_cast<VirtualQAbstractFileIconProvider::QAbstractFileIconProvider_Options_Callback>(slot));
}
}

void QAbstractFileIconProvider_Delete(QAbstractFileIconProvider* self) {
    delete self;
}

